core/translate/planner.rs

```diff
@@ -494,6 +484,49 @@ pub fn prepare_select_plan<'a>(schema: &Schema, select: ast::Select) -> Result<P
     }
 }
 
+pub fn prepare_delete_plan(
+    schema: &Schema,
+    tbl_name: &QualifiedName,
+    where_clause: Option<Expr>,
+) -> Result<Plan> {
+    let table_name = tbl_name.name.0.clone();
+
+    let table = if let Some(table) = schema.get_table(&table_name) {
+        table
+    } else {
+        crate::bail_parse_error!("Table {} not found", table_name);
+    };
+
+    let table_ref = BTreeTableReference {
+        table: table.clone(),
+        table_identifier: table_name.clone(),
+        table_index: 0,
+    };
+
+    // Parse and resolve the where_clause
+    let resolved_where_clauses = parse_where(where_clause, &[table_ref.clone()])?;
+
+    let plan = Plan {
```

Comment:
There's a lot of fields on `Plan` that a DELETE operation does not need (or cannot have). Wondering if we should just have:

```
struct SelectPlan {...}
struct DeletePlan {...}

enum Plan {
  Select(SelectPlan),
  Delete(DeletePlan)
}
```

so you could call `emit_program(plan: Plan)` and it could pattern match on the plan type to decide what to do. not immediately sure if that makes the code better or worse but you could try and see if you have time and energy

Reply:
Pros:

- Exposes a single `optimize_plan` and `emit_program` function, allowing pattern matching within each module.
- `DeletePlan` contains only the required fields.

Cons:

- Most fields in `SelectPlan` and `DeletePlan` are duplicated. However, I prefer to have duplicates rather than using a single struct where we must remember which fields belong to select or delete.

As I think there are more pros, hence chose to split plan into `Select` and `Delete`